libcasr/src/sarif.rs

```diff
@@ -0,0 +1,404 @@
+            "problem.severity".to_string(),
+            Value::String(class.severity.clone()),
+        );
+        rule.insert("properties".to_string(), Value::Object(properties));
+
+        (Some(Value::Object(rule)), rule_id)
+    }
+
+    /// Add CASR CrashReport to SARIF report.
+    /// NOTE: before using this method,
+    /// use SarifReprot::new() to get report.
+    ///
+    ///  # Arguments
+    ///
+    ///  * 'report' - CrashReport
+    ///
+    ///  * 'source_root' - Path to source root directory.
+    pub fn add_casr_report<T: AsRef<Path>>(
+        &mut self,
+        report: &CrashReport,
+        source_root: T,
+    ) -> Result<()> {
+        let (rule, rule_id) = self.rule(&report.execution_class);
+        if let Some(rule) = rule {
+            let rules = self.json.as_object_mut().unwrap()["runs"]
+                .as_array_mut()
+                .unwrap()[0]
+                .as_object_mut()
+                .unwrap()["tool"]
+                .as_object_mut()
+                .unwrap()["driver"]
+                .as_object_mut()
+                .unwrap()["rules"]
+                .as_array_mut()
+                .unwrap();
+            rules.push(rule);
+            if rule_id.starts_with('G') {
+                self.id += 1;
+            }
+        }
+
+        let results = self.json.as_object_mut().unwrap()["runs"]
+            .as_array_mut()
+            .unwrap()[0]
+            .as_object_mut()
+            .unwrap()["results"]
+            .as_array_mut()
+            .unwrap();
+        let mut result = Map::new();
+        result.insert("ruleId".to_string(), Value::String(rule_id));
+        result.insert("level".to_string(), Value::String("error".to_string()));
+        let mut message = Map::new();
+        let text = if !report.stdin.is_empty() {
+            format!(
+                "{}: {} < {}",
+                report.execution_class.short_description, report.proc_cmdline, report.stdin
+            )
+        } else {
+            format!(
+                "{}: {}",
+                report.execution_class.short_description, report.proc_cmdline
+            )
+        };
+        message.insert("text".to_string(), Value::String(text));
+        result.insert("message".to_string(), Value::Object(message));
+        let mut locations: Vec<Value> = Vec::new();
+        let mut location = Map::new();
+        let mut physical_loc = Map::new();
+        let mut artifact_loc = Map::new();
+        let mut region = Map::new();
+        let parts: Vec<_> = report.crashline.split(':').map(|s| s.to_string()).collect();
+        if parts.len() != 2 && parts.len() != 3 {
+            return Err(Error::Casr(format!(
+                "Unable to parse crashline: {}",
+                report.crashline
+            )));
+        }
+
+        let source_path = Path::new(&parts[0]);
+        let norm_source_path =
+            if let Ok(norm_source_path) = source_path.strip_prefix(source_root.as_ref()) {
```

Comment:
We should make both paths canonical before stripping prefix. Consider the following example:

```
source_root = /my/src/dir
source_path = /my/../my/src/dir
```

Reply:
I saw these real examples in stack traces

Reply:
Hm, this is the problem, because. We couldn't use [std](https://doc.rust-lang.org/stable/std/index.html)::[fs](https://doc.rust-lang.org/stable/std/fs/index.html)::[canonicalize](https://doc.rust-lang.org/stable/std/fs/fn.canonicalize.html#), because source path could exist. It is an open question what to do.

Reply:
https://docs.rs/lexiclean/latest/lexiclean/